utils/nctl/sh/scenarios/itst02.sh

```diff
@@ -0,0 +1,165 @@
+    sleep_and_compare "60"
+    # 6-8. Restart three nodes
+    do_start_node "5"
+    do_start_node "4"
+    do_start_node "3"
+    # 9-11. Check sync of restarted node
+    do_await_full_synchronization "5"
+    do_await_full_synchronization "4"
+    do_await_full_synchronization "3"
+    # 12. Ensure era proceeds after restart
+    do_await_era_change "2"
+    # 13-15. Check sync of nodes again
+    do_await_full_synchronization "5"
+    do_await_full_synchronization "4"
+    do_await_full_synchronization "3"
+    # 16-18. Compare stalled lfb hash to current
+    compare "5" "$STALLED_LFB"
+    compare "4" "$STALLED_LFB"
+    compare "3" "$STALLED_LFB"
+
+    log "------------------------------------------------------------"
+    log "Scenario itst02 complete"
+    log "------------------------------------------------------------"
+}
+
+function log_step() {
+    local COMMENT=${1}
+    log "------------------------------------------------------------"
+    log "STEP $STEP: $COMMENT"
+    STEP=$((STEP + 1))
+}
+
+function compare() {
+    log_step "node-${1}: checking chain progressed"
+    local LFB1=$(do_read_lfb_hash ${1})
+    local LFB2=${2}
+
+    if [ "$LFB1" = "$LFB2" ]; then
+       log "error: $LFB1 = $LFB2, chain didn't progress."
+       exit 1
+    fi
+}
+
+function sleep_and_compare() {
+    log_step "ensuring chain stalled"
+    local SLEEP_TIME=${1}
+    local LFB1=$(do_read_lfb_hash 1)
+    log "Sleeping ${SLEEP_TIME}s..."
+    sleep $SLEEP_TIME
+    local LFB2=$(do_read_lfb_hash 1)
+
+    if [ "$LFB1" != "$LFB2" ]; then
+       log "Error: Chain progressed."
+       exit 1
+    fi
+
+    STALLED_LFB=$LFB2
+}
+
+function do_await_genesis_era_to_complete() {
+    log_step "awaiting genesis era to complete"
+    while [ "$(get_chain_era)" -lt 1 ]; do
+        sleep 1.0
+    done
+}
+
+function do_read_lfb_hash() {
+    local NODE_ID=${1}
+    LFB_HASH=$(render_last_finalized_block_hash "$NODE_ID" | cut -f2 -d= | cut -f2 -d ' ')
+    echo "$LFB_HASH"
+}
+
+function do_stop_node() {
+    local NODE_ID=${1}
+    log_step "stopping node-$NODE_ID."
+    do_node_stop "$NODE_ID"
+    sleep 1
+}
+
+function do_start_node() {
+    local NODE_ID=${1}
+    log_step "starting node-$NODE_ID. Syncing from hash=${STALLED_LFB}"
+    do_node_start "$NODE_ID" "$STALLED_LFB"
+    sleep 1
+    if [ "$(do_node_status ${NODE_ID} | awk '{ print $2 }')" != "RUNNING" ]; then
+        log "ERROR: node-${NODE_ID} is not running"
+	exit 1
+    fi
+}
+
+function do_await_full_synchronization() {
+    local NODE_ID=${1}
+    local WAIT_TIME_SEC=0
+    log_step "awaiting full synchronization of node=${NODE_ID}…"
+    while [ "$(do_read_lfb_hash "$NODE_ID")" != "$(do_read_lfb_hash "1")" ]; do
+        if [ "$WAIT_TIME_SEC" = "$SYNC_TIMEOUT_SEC" ]; then
+            log "ERROR: Failed to synchronize node-${NODE_ID} in ${SYNC_TIMEOUT_SEC} seconds"
+            exit 1
+        fi
+        WAIT_TIME_SEC=$((WAIT_TIME_SEC + 1))
+        sleep 1.0
+    done
+}
```

Comment:
We should have the double assertion that we have in the other tests. Otherwise, we might be getting false positives and ignore a case where the node after syncing does not transition to an active participant of the protocol. I know it's slower but it's the only way we can be sure it's working correctly :grimacing:

Reply:
I actually do test it twice, I just rearranged how.

`# 9-11. Check sync of restarted node
    do_await_full_synchronization "5"
    do_await_full_synchronization "4"
    do_await_full_synchronization "3"
    # 12. Ensure era proceeds after restart
    do_await_era_change "2"
    # 13-15. Check sync of nodes again
    do_await_full_synchronization "5"
    do_await_full_synchronization "4"
    do_await_full_synchronization "3"`

Reply:
Ah, I see. I missed that :) We should write this test function that asserts that LFB is the same across all nodes.

Reply:
added a fuction to check across all of the nodes used in this test: https://github.com/CasperLabs/casper-node/pull/959/files#diff-8bd21e6750b7829a78940cd1d0405bfe8e4ba19e78acb53c085cb8cfba164d41R138

I call it prior to stopping the nodes (https://github.com/CasperLabs/casper-node/pull/959/files#diff-8bd21e6750b7829a78940cd1d0405bfe8e4ba19e78acb53c085cb8cfba164d41R29) and after we check that each individual node starts up (https://github.com/CasperLabs/casper-node/pull/959/files#diff-8bd21e6750b7829a78940cd1d0405bfe8e4ba19e78acb53c085cb8cfba164d41R51)